### ISSUE TITLE
Rule id-blacklist incorrectly flags object property reassignments

### ISSUE THREAD

**Issue by MaffooBristol** [eslint/eslint#12787]
**Tell us about your environment**

* **ESLint Version:** 6.8.0
* **Node Version:** 10.15.3
* **npm Version:** 6.11.3

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```yaml
extends: airbnb
parser: babel-eslint
rules:
  space-before-function-paren: 2
  brace-style: [2, 'stroustrup']
  arrow-body-style: 0
  no-underscore-dangle: 0
  class-methods-use-this: 1
  consistent-return: 0
  no-unused-vars: 1
  no-bitwise: 0
  no-plusplus: 0
  linebreak-style: 0
  arrow-parens: [2, 'always']
  no-useless-escape: 0
  curly: [2, 'all']
  no-lonely-if: 1
  semi: 2
  max-len: [1, { code: 120, ignoreComments: true }]
  prefer-destructuring: 1
  lines-between-class-members: 0
  object-curly-newline: 0
  indent: 1
  jsx-a11y/anchor-is-valid: [ 'error', {
    components: [ 'Link' ],
    specialLink: [ 'to' ]
  }]
  jsx-a11y/click-events-have-key-events: 0
  jsx-a11y/no-static-element-interactions: 0
  jsx-a11y/no-noninteractive-tabindex: 0
  react/prefer-stateless-function: 1
  react/no-unused-state: 1
  react/state-in-constructor: [1, "never"]
  react/jsx-one-expression-per-line: 0
  react/prop-types: 0 # This should be enabled at some point
  react/jsx-indent: 1
  import/no-extraneous-dependencies: ["error", {"devDependencies": ["test/**/*.js"]}]
  jsx-a11y/label-has-associated-control: [ 2, {
    depth: 1,
  }]
  react/forbid-component-props: ["error", {"forbid": ["style", "className"]}]
  react/forbid-dom-props: ["error", {"forbid": ["style", "className"]}]
  react/jsx-sort-props: ["error", {"callbacksLast": true, "reservedFirst": true, "shorthandFirst": true, "noSortAlphabetically": true}]
  id-blacklist: [1, "data", "err", "e", "cb", "callback"]
env:
  browser: true
  node: true
  es6: true
```

</details>

Highlight:
```yaml
id-blacklist: [1, "data", "err", "e", "cb", "callback"]
```

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
const { data: user } = await axios.post('/api/auth', loginCredentials);
if (!user) {
  throw new Error('Error fetching authentication data');
}
```

```bash
npx eslint --ext .js --ext .jsx src/
```

**What did you expect to happen?**
The `data` property shouldn't be flagged here because it's coming _from_ the axios request and there's nothing I can do about it. I'm converting the value `response.data` to the variable `user`. However it highlights it as being as if I were assigning the data variable.

**What actually happened? Please include the actual, raw output from ESLint.**
```
45:15  warning  Identifier 'data' is blacklisted  id-blacklist
```

**Are you willing to submit a pull request to fix this bug?**
I would but I'm not sure how as I've never delved into the eslint source!

**Comment by MaffooBristol** [comment]
Equivalent code that doesn't flag anything but does the same thing:

```javascript
const authResponse = await axios.post('/api/auth', loginCredentials);
const { user } = authResponse.data;
```

**Comment by MaffooBristol** [comment]
Additionally it has issues with optional chaining syntax:

```javascript
if (error.response?.data?.error) { // Warns use of data
  return error.response.data.error;
}
```

**Comment by aladdin-add** [comment]
it seems a bug to me, as the docs say:

> 
> It will not catch blacklisted identifiers that are:
> * function calls (so you can still use functions you do not have control over)
> * object properties (so you can still use objects you do not have control over)

re optional chaining, please follow https://github.com/eslint/eslint/issues/12642

**Comment by kaicataldo** [comment]
I am able to reproduce this using our [demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IGlkLWJsYWNrbGlzdDogWzEsIFwiZGF0YVwiLCBcImVyclwiLCBcImVcIiwgXCJjYlwiLCBcImNhbGxiYWNrXCJdICovXG5cbmFzeW5jIGZ1bmN0aW9uIGZvbygpIHtcbiAgY29uc3QgeyBkYXRhOiB1c2VyIH0gPSBhd2FpdCBheGlvcy5wb3N0KCcvYXBpL2F1dGgnLCBsb2dpbkNyZWRlbnRpYWxzKTtcbiAgaWYgKCF1c2VyKSB7XG4gICAgdGhyb3cgbmV3IEVycm9yKCdFcnJvciBmZXRjaGluZyBhdXRoZW50aWNhdGlvbiBkYXRhJyk7XG4gIH1cbn0iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjExLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7ImpzeCI6dHJ1ZX19LCJydWxlcyI6e30sImVudiI6eyJicm93c2VyIjpmYWxzZX19fQ==).

This seems doubly broken to me, since the the identifier is being renamed. We should update this rule to not warn on object destructured identifiers.
